### ISSUE TITLE
Gdiff: any BufWritePost autocommand disables diff mode when writing index buffer

### ISSUE THREAD

**Issue by smancill** [tpope/vim-fugitive#1019]
How to reproduce:

1. Open a modified file and run `Gdiff` (see below for contents of `minimal.vim`):

        vim -u minimal.vim -U none -i none +Gdiff vimrc

2. The diff looks fine:

    <img width="888" alt="captura de pantalla 2018-04-17 a la s 20 54 52" src="https://user-images.githubusercontent.com/238528/38904709-71102302-4282-11e8-8aaf-3326b5304fb5.png">

3. But now write the *index* buffer (`:w`), the diff mode is gone:

    <img width="891" alt="captura de pantalla 2018-04-17 a la s 20 55 05" src="https://user-images.githubusercontent.com/238528/38904782-c96c627c-4282-11e8-8ff6-68a07566638f.png">

    Both buffers now get out of sync when scrolling, the diff syntax acts all weird, etc.
    
4. The diff mode can be restored by running `:diffthis` in the *index* buffer after the `:w`.

I have tracked the problem to any user defined `BufWritePost` autocommand running when the fugitive buffer is saved.

`minimal.vim` (using [vim-plug](https://github.com/junegunn/vim-plug)):
```vim
set nocompatible

call plug#begin('~/.vim/bundle')

Plug 'tpope/vim-sensible'
Plug 'tpope/vim-fugitive'

augroup minimal
autocmd!
autocmd BufWritePost * echomsg "New BufWritePost autocmd"
augroup end

call plug#end()
```

If the autocommand is removed, then everything works fine.

Vim version:
```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Apr 13 2018 06:21:17)
macOS version
Included patches: 1-1700
Compiled by Homebrew
Huge version without GUI.  Features included (+) or not (-):
+acl               +farsi             +mouse_sgr         -tag_any_white
+arabic            +file_in_path      -mouse_sysmouse    -tcl
+autocmd           +find_in_path      +mouse_urxvt       +termguicolors
-autoservername    +float             +mouse_xterm       +terminal
-balloon_eval      +folding           +multi_byte        +terminfo
+balloon_eval_term -footer            +multi_lang        +termresponse
-browse            +fork()            -mzscheme          +textobjects
++builtin_terms    -gettext           +netbeans_intg     +timers
+byte_offset       -hangul_input      +num64             +title
+channel           +iconv             +packages          -toolbar
+cindent           +insert_expand     +path_extra        +user_commands
-clientserver      +job               +perl              +vertsplit
+clipboard         +jumplist          +persistent_undo   +virtualedit
+cmdline_compl     +keymap            +postscript        +visual
+cmdline_hist      +lambda            +printer           +visualextra
+cmdline_info      +langmap           +profile           +viminfo
+comments          +libcall           -python            +vreplace
+conceal           +linebreak         +python3           +wildignore
+cryptv            +lispindent        +quickfix          +wildmenu
+cscope            +listcmds          +reltime           +windows
+cursorbind        +localmap          +rightleft         +writebackup
+cursorshape       -lua               +ruby              -X11
+dialog_con        +menu              +scrollbind        -xfontset
+diff              +mksession         +signs             -xim
+digraphs          +modify_fname      +smartindent       -xpm
-dnd               +mouse             +startuptime       -xsmp
-ebcdic            -mouseshape        +statusline        -xterm_clipboard
+emacs_tags        +mouse_dec         -sun_workshop      -xterm_save
+eval              -mouse_gpm         +syntax            
+ex_extra          -mouse_jsbterm     +tag_binary        
+extra_search      +mouse_netterm     +tag_old_static    
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: clang -c -I. -Iproto -DHAVE_CONFIG_H   -DMACOS_X -DMACOS_X_DARWIN  -g -O2 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1       
Linking: clang   -L. -fstack-protector -L/usr/local/lib -L/usr/local/opt/libyaml/lib -L/usr/local/opt/openssl/lib -L/usr/local/opt/readline/lib  -L/usr/local/lib -o vim        -lncurses -liconv -framework AppKit   -mmacosx-version-min=10.12 -fstack-protector-strong -L/usr/local/lib  -L/usr/local/Cellar/perl/5.26.1/lib/perl5/5.26.1/darwin-thread-multi-2level/CORE -lperl -lm -lutil -lc  -L/usr/local/opt/python/Frameworks/Python.framework/Versions/3.6/lib/python3.6/config-3.6m-darwin -lpython3.6m -framework CoreFoundation  -lruby.2.5.1 -lobjc    
```

Fugitive version:
```
v2.2-129-g40d78f07dee2
```

Git version:
```
git version 2.17.0
```

System version:
```
ProductName:	Mac OS X
ProductVersion:	10.12.4
BuildVersion:	16E195
```

I am not sure if the problem is with Fugitive or introduced by a Vim update. Vim 7.4 (the system version distributed with macOS) works fine.

**Comment by smancill** [comment]
Fixed by one of the patches between Vim 8.0.1700 and 8.0.1750 (after updating the Homebrew version).

**Comment by tpope** [comment]
Good to know, thanks.
